### PR TITLE
[4.1] Fixing "#__finder_tokens full" error

### PR DESCRIPTION
**Pull request by Hackwar**
### Summary of Changes
Smart Search/Finder has a long standing issue that the #__finder_tokens table can run full and abort the indexing process. That table is of type "MEMORY", which can run full and there is an option to reduce the number of tokens to store in that table, which unfortunately doesn't work in all cases. The MEMORY table is used since it is a lot faster than a normal InnoDB table and since this is just used for temporary storage, writing speed is the main factor here. Anyway, enough rambling. ;) This change pushes the code of `addTokenToDb` to the `tokenizeToDbShort()` method and properly counts the number of tokens written to DB. This means, that now the option mentioned above works properly.
This removes the `addTokenToDb` method, but since this is a protected method in a class of a component, this would still be in accordance to our backwards compatibility promise.

### Testing Instructions
- Get a very long text and insert it into an article.
- save it.

### Actual result BEFORE applying this Pull Request
The indexing fails.

### Expected result AFTER applying this Pull Request
The indexing works.

### CODE REVIEW

**Comment by richard67** [comment]
@Hackwar I have tested this PR but it doesn't change anything for me. I get the error with and without the PR applied when trying to save an article with some 100 or 200 times the text from the site linked in the description of issue #34377 , https://whiletrue.neocities.org/lte.html  .

**Comment by bembelimen** [comment]
Just playing around with this thing: The problem of the whole thing is most likely that the number of the ```memory_table_limit``` parameter is too high + the check is on the wrong position.

So first the checks should be enough to have it that way: [patch.zip](https://github.com/joomla/joomla-cms/files/7591586/patch.zip)

And second either set the parameter to a lower number (works for me with 20.000) or take the number from https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_heap_table_size

**Comment by Krshivam25** [comment]
error

File: administrator/components/com_finder/src/Indexer/Indexer.php [hunk 1]
```diff
@@ -951,31 +917,31 @@ protected function addTokensToDb($tokens, $context = '')
...
-					. $db->quote($token->stem) . ', '
+					$query->quote($token->term) . ', '
+					. $query->quote($token->stem) . ', '
 					. (int) $token->common . ', '
 					. (int) $token->phrase . ', '
-					. $db->quote($token->weight) . ', '
+					. $query->quote($token->weight) . ', '
 					. (int) $context . ', '
-					. $db->quote($token->language)
+					. $query->quote($token->language)
```
Comment (HLeithner) [inline administrator/components/com_finder/src/Indexer/Indexer.php:926]: What's the reason to make it slower then needed? please use the `quote` function of the database as before, the `quote` function in the Query object is only a proxy, which is an overhead.

**Comment by bembelimen** [comment]
It seems, this will not make it in 4.1.0, so I moved it to 4.1.1 to not forget it, as it should be fixed asap.

**Comment by Hackwar** [comment]
Closing this PR in favour of #36749

**Pull request closed without merging**